frontend/public/components/utils/value-from-pair.jsx

```diff
@@ -22,21 +26,102 @@ import * as PropTypes from 'prop-types';
 //       resource: requests.cpu
 //       divisor: 1 // 1 is default
 
-const FieldRef = ({data: {fieldPath}}) => <span>
+const getSpacer = (configMap, secret) => {
+  const spacerBefore = new Set();
+  return _.isEmpty(configMap) ? spacerBefore : spacerBefore.add(secret);
+};
+
+const getHeaders = (configMap, secret) => {
+  return _.isEmpty(configMap) ? {
+    [secret]: 'Secrets'
+  } : {
+    [configMap]: 'Config Maps',
+    [secret]: 'Secrets'
+  };
+};
 
+const getKeys = (itemKeys, keyMap) => {
+  Object.keys(keyMap).map((value) => {
+    itemKeys[value] = value;
+  });
+  return itemKeys;
+};
+
+const NameKeyDropdownPair = ({name, key, configMaps, secrets, onChange, kind, nameTitle, placeholderString}) => {
+  let itemKeys = {};
+  let keyRefString;
+  const cmItems = {};
+  const secretItems = {};
+  const nameAutocompleteFilter = (text, item) => fuzzy(text, item.props.name);
+  const keyAutocompleteFilter = (text, item) => fuzzy(text, item);
+  const keyTitle = _.isEmpty(key) ? <span className="text-muted">Select a key</span> : <span>{key}</span>;
+
+  if (configMaps && configMaps.items) {
+    _.each(configMaps.items, (v) => {
+      cmItems[`${v.metadata.name}:configMapKeyRef`] = <ResourceName kind="ConfigMap" name={v.metadata.name}/>;
+      if (kind === 'ConfigMap' && _.isEqual(v.metadata.name, name)) {
+        keyRefString = 'configMapKeyRef';
+        itemKeys = getKeys(itemKeys, v.data);
+      }
+    });
+  }
+  if (secrets && secrets.items) {
+    _.each(secrets.items, (v) => {
+      secretItems[`${v.metadata.name}:secretKeyRef`] = <ResourceName kind="Secret" name={v.metadata.name}/>;
+      if (kind === 'Secret' && _.isEqual(v.metadata.name, name)) {
+        keyRefString = 'secretKeyRef';
+        itemKeys = getKeys(itemKeys, v.data);
+      }
+    });
+  }
+  const firstConfigMap = _.isEmpty(cmItems) ? {} : Object.keys(cmItems)[0];
+  const firstSecret = _.isEmpty(secretItems) ? {} : Object.keys(secretItems)[0];
+  const headerBefore = getHeaders(firstConfigMap, firstSecret);
+  const spacerBefore = getSpacer(firstConfigMap, firstSecret);
+  const items = _.assign({}, cmItems, secretItems);
+  return <span>
+    <Dropdown
+      menuClassName="co-namespace-selector__menu"
+      className="value-from-pair"
+      autocompleteFilter={nameAutocompleteFilter}
+      autocompletePlaceholder={placeholderString}
+      items={items}
+      selectedKey={name}
+      title={nameTitle}
+      headerBefore={headerBefore}
+      spacerBefore={spacerBefore}
+      onChange={val => onChange({[_.split(val, ':')[1]]:{'name': _.split(val, ':')[0], 'key': ''}})}
```

Comment:
I feel like there's too much logic inline here.

Reply:
Sure thing. Broke out the split and made a more legible block of code.